### PR TITLE
model: support gemma4

### PR DESCRIPTION
**Pull request by rebel-eunji**
### 🚀 Summary of Changes

- padding logic is different with Gemma3. Gemma4 calculated the padding size with variable image size.
- It is not supported current dev. It can be run after dev-0.22.0 branch is merged into dev.

---

### 📌 Related Issues / Tickets

* Resolves #
* Related to #

---

### ✅ Type of Change

* [ ] 🚀 Release (`release`)
* [ ] ✨ Feature (`feature`)
* [ ] 🧠 Model support (`model`)
* [ ] 🧬 Core engine changes (`core`)
* [ ] 🛠 Bug fix (`fix`)
* [ ] ⚙️ Performance improvement (`perf`)
* [ ] 🔁 Refactor or code cleanup (`refactor`)
* [ ] 📄 Documentation (`docs`)
* [ ] ❓ Other (`other`): please describe

---

### 🧪 How to Test

1. Run `...`
2. Verify output: `...`
3. Edge case tested: `...`

---

### 📸 Screenshots / Logs (if applicable)

---

### 📋 Checklist

* [ ] PR title follows Conventional Commits format
* [ ] This PR is linked to an existing issue
* [ ] The test method is described, and the expected result is clearly stated
* [ ] Relevant documentation has been updated (if applicable)

---

### 💬 Notes

---

### CODE REVIEW

**Comment by rebel-jonghewk** [comment]
@rebel-eunji is this ready for review?

**Comment by rebel-eunji** [comment]
@rebel-jonghewk I think this PR should be merged after optimum-rbln's PR is in. https://github.com/RBLN-SW/optimum-rbln/pull/563.

@rebel-dkhong When do you plan to merge your PR?

**Comment by rebel-eunji** [comment]
@rebel-jonghewk This PR is almost similar with gemma3. And then the code will be refactored with other multimodal models including gemma3.

File: vllm_rbln/model_executor/models/optimum/gemma4.py [hunk 1]
```diff
@@ -0,0 +1,352 @@
+# Copyright 2025 Rebellions Inc. All rights reserved.
```
Comment (rebel-jonghewk) [inline vllm_rbln/model_executor/models/optimum/gemma4.py:1]: Can we combine some of the logics with gemma3.py?

File: vllm_rbln/model_executor/models/optimum/gemma4.py [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+        # not image next") and assert the fixed length. Remove this guard once the
+        # dynamic per-image token count above is supported.
+        EXPECTED_IMAGE_TOKENS = 280
+        next_is_image = torch.cat(
+            [token_type_ids[1:], token_type_ids.new_zeros(size=(1,))]
+        )
+        image_ends = torch.where(token_type_ids & ~next_is_image)[0]
+        block_lengths = image_ends - image_starts + 1
+        assert torch.all(block_lengths < EXPECTED_IMAGE_TOKENS), (
+            f"Expected each image block to be {EXPECTED_IMAGE_TOKENS} tokens, "
```
Comment (rebel-jonghewk) [inline vllm_rbln/model_executor/models/optimum/gemma4.py]: nit: smaller than?

File: vllm_rbln/model_executor/models/optimum/gemma4.py [hunk 3]
```diff
@@ -0,0 +1,352 @@
...
+            #     mm_input_by_modality["video"] = self._parse_and_validate_video_input(
+            #         **kwargs
+            #     )
+            # if (
+            #     input_key == "input_features_padded"
+            #     and "audio" not in mm_input_by_modality
+            # ):
+            #     mm_input_by_modality["audio"] = self._parse_and_validate_audio_input(
+            #         **kwargs
+            #     )
```
Comment (rebel-jonghewk) [inline vllm_rbln/model_executor/models/optimum/gemma4.py]: nit: erase?

File: vllm_rbln/model_executor/models/optimum/gemma3.py [hunk 4]
```diff
@@ -215,6 +214,21 @@ def forward(self, model_input: ModelInputForRBLN, **kwargs) -> torch.Tensor:
             logits = logits[:request_nums]
         return logits
 
+    def _build_prefill_embeds(
```
Comment (rebel-eunji) [inline vllm_rbln/model_executor/models/optimum/gemma3.py:217]: This code will be refactored in bump refactoring PR.
https://github.com/RBLN-SW/vllm-rbln/pull/664

File: vllm_rbln/model_executor/models/optimum/gemma4.py (hunk 1 above)
Comment (rebel-eunji) [reply vllm_rbln/model_executor/models/optimum/gemma4.py:1]: I've removed duplicated codes.

File: vllm_rbln/model_executor/models/optimum/gemma4.py (hunk 3 above)
Comment (rebel-eunji) [reply vllm_rbln/model_executor/models/optimum/gemma4.py]: https://github.com/RBLN-SW/vllm-rbln/pull/622/changes/2bc25b16c4ec42448198ef4d84644602585a63c8

File: vllm_rbln/model_executor/models/optimum/gemma4.py (hunk 2 above)
Comment (rebel-eunji) [reply vllm_rbln/model_executor/models/optimum/gemma4.py]: Currently, only a maximum of 280 image tokens is supported.
cc. @rebel-shshin

Comment (rebel-jonghewk) [reply vllm_rbln/model_executor/models/optimum/gemma4.py]: "Expected each image block to be {EXPECTED_IMAGE_TOKENS} tokens, " -> Then I think you need to change the error message

Comment (rebel-eunji) [reply vllm_rbln/model_executor/models/optimum/gemma4.py]: https://github.com/RBLN-SW/vllm-rbln/pull/622/changes/c80ee3fed04239cae96ce7506e68f57bdb791a26

**Review by rebel-jonghewk — APPROVED** [review]
